### ISSUE TITLE
Enable custom reporters

### ISSUE THREAD

**Issue by bradwilson** [xunit/xunit#1874]
Today, end users cannot write custom reporters because they must link against explicit versions of runner.utility, and therefore would only be usable by runners linked against the same version of runner utility (or risk runtime failures related to types).

Let's see if there's anything we can do about that in 3.0.

**Comment by bradwilson** [comment]
## Implementation

The implementation takes advantage of the new v3 architecture where every test project is a stand-alone executable. Runner reporters will get linked into the test project, and registered via an assembly-level attribute (reporters distributed via NuGet packages may choose to do automatically on behalf of the user).

As with v2, you will be expected to provide an implementation of `IRunnerReporter` and an associated `IRunnerReporterMessageHandler`. The type which implements `IRunnerReporter` is registered via an assembly-level attribute (example in C#):

```csharp
[assembly: RegisterRunnerReporter(typeof(MyReporter))]
```

There are two new properties on `IRunnerReporter` compared to v2:

```csharp
/// <summary>
/// Gets a value which indicates if it's possible for this reporter to be environmentally
/// enabled.
/// </summary>
/// <remarks>
/// Note that this differs from <see cref="IsEnvironmentallyEnabled"/> which checks to see whether
/// the conditions currently exist to environmentally enable the reporter. This value is used when
/// constructing the console runner help output that lists which runners might be environmentally
/// enabled.
/// </remarks>
bool CanBeEnvironmentallyEnabled { get; }
```

and

```csharp
/// <summary>
/// Gets a value which indicates whether this runner wishes to force no logo.
/// Useful for runners which are designed for purely parseable output
/// (for example, <see cref="JsonReporter"/>).
/// </summary>
bool ForceNoLogo { get; }
```

## Command line

Specifying the runner reporter is done via `-reporter <switch>` on the command line.

The command line help has been updated (shown here with the built-in reporters):

```
Runner reporters (optional, choose only one)

  -reporter <option> : choose a reporter
                     :   default  - show standard progress messages
                     :   json     - show full progress messages in JSON [implies '-noLogo']
                     :   quiet    - only show failure messages
                     :   silent   - do not show any messages [implies '-noLogo']
                     :   teamCity - TeamCity CI support
                     :   verbose  - show verbose progress messages

  The following reporters will be automatically enabled in the appropriate environment.
  An automatically enabled reporter will override a manually selected reporter.
    Note: You can disable auto-enabled reporters by specifying the '-noAutoReporters' switch

    * AppVeyor CI support
    * Azure DevOps/VSTS CI support
    * TeamCity CI support
```

_**Note:** The old command line switches (`-json`, `-quiet`, `-silent`, `-teamCity`, and `-verbose`) are supported for backward compatibility, but do not show up in the help output._

_**Note:** reporter choices are ignored in `-automated` mode; automated mode always uses the built-in JSON reporter._

## Built-in reporters

The built-in runner reporters will use this same registration mechanism, and you can opt-out of having any of the built-in reporters registered by default, by specifying the following in your project file (`.csproj`/`.fsproj`/`.vbproj`):

```xml
<PropertyGroup>
  <XunitRegisterBuiltInRunnerReporters>false</XunitRegisterBuiltInRunnerReporters>
</PropertyGroup>
```

By allowing developers to remove the built-in runner reporters, they can control the exact reporters that will be available (including providing new custom reporters that can use the existing switches). This includes replacing the `default` reporter, which is what will be used when no reporter has been chosen. (If no runner reporter has been registered with the `default` switch, then the system will fall back to using the built-in `DefaultRunnerReporter` class.)

## Multi-assembly runners (like `xunit.v3.runner.console`)

Multi-assembly runners can only use the built-in runner reporters. The only way to get custom runner reporters is to directly run your test project.

The same registration mechanism is used with the multi-assembly runners, except they're registered into the runner assembly rather than the test assembly. Advanced developers could pull the source for xUnit.net and modify the default runner reporter list, and rebuild the multi-assembly runners, so that the multi-assembly runners would also have access to custom runner reporters. _**There are no plans for supporting v2-style external runner reporters in v3.**_

**Comment by bradwilson** [comment]
Available in v3 `0.5.0-pre.18` https://xunit.net/docs/using-ci-builds
